### PR TITLE
Dense no cython extra functions

### PR DESCRIPTION
**Pull request by MrRobot2211**
This PR will contain the extra function of the `CuPyDense` class which are not part of QuTiP's `data.Data`

### CODE REVIEW

**Comment by hodgestar** [comment]
@MrRobot2211 I'm going to hold off reviewing this until #4 is merged or closer to merging, but please feel free to continue here as you like. :D

**Comment by hodgestar** [comment]
@MrRobot2211 One #19 is merged, could you merge main into this branch and fix up any things flake8 or black complain about and then ping me when you're ready for this to be reviewed?

**Comment by MrRobot2211** [comment]
> @MrRobot2211 Once #19 is merged, could you merge main into this branch and fix up any things flake8 or black complain about and then ping me when you're ready for this to be reviewed?

 
Yes, of course @hodgestar. There are many functions missing here, as well.  After I merge the changes and complete all the functions  I'll ping you.

**Comment by MrRobot2211** [comment]
Well #15 is already handled.

File: src/qutip_cupy/dense.py
Comment (hodgestar) [inline src/qutip_cupy/dense.py:130]:
```suggestion
```
Remove blank line.

File: src/qutip_cupy/dense.py
Comment (hodgestar) [inline src/qutip_cupy/dense.py:141]:
```suggestion
```
We can remove `__dealloc__` -- this class is pure Python.

File: src/qutip_cupy/dense.py
Comment (hodgestar) [inline src/qutip_cupy/dense.py:161]: I suggest we remove `empty` and `empty_like` from this PR, and then add create an issue in this repo to add them and to add the dispatchers to qutip dev.major.

File: src/qutip_cupy/dense.py
Comment (hodgestar) [inline src/qutip_cupy/dense.py:178]:
```suggestion
    if scale == 1:
```
We should also add tests for `scale=1`, scale=`2`, scale=`1+5j`.

File: src/qutip_cupy/dense.py
Comment (hodgestar) [inline src/qutip_cupy/dense.py:189]:
```suggestion
```
Added the suggestion to remove the empty function.

File: src/qutip_cupy/dense.py
Comment (hodgestar) [inline src/qutip_cupy/dense.py:324]: E302 is "Expected two blank lines". Why do we ignore that here? It looks like there are two blank lines above this function definition? Same question for other functions.

**Review by hodgestar — COMMENTED** [review]
I know this is still in progress, but I did a partial review of what is here to give some earlier feedback and hopefully save some time later.

Broader comments:
* We need a lot more tests (probably a few for each new specialization function)
* The implementation of `diags` looks very odd in places, but I assume it's a work in progress.

File: src/qutip_cupy/dense.py
Comment (MrRobot2211) [reply src/qutip_cupy/dense.py:324]: I might have had a mix-up here, but I thought that we inclined to having smaller spacing of short funtions, so that the files do not become to long and is eadier for reading.

File: src/qutip_cupy/__init__.py
Comment (MrRobot2211) [reply src/qutip_cupy/__init__.py:52]: I made the same question on issue #15 see [this](f.append(experiment_dict) I am sticking to the naming used in qutip dev.major, but I think the proper thing would be to extend the naming convention as I suggest there,

File: src/qutip_cupy/dense.py
Comment (MrRobot2211) [reply src/qutip_cupy/dense.py:161]: When you say the dispatchers  you refer to  `empty`'s and `empty-like`'s dispatcher right? 

Shouldn't we move all the constructors to a separate PR then?

File: src/qutip_cupy/__init__.py
Comment (MrRobot2211) [reply src/qutip_cupy/__init__.py:52]: In fact `diags` is the name given in `dev.major`  but I am happier with `diag`.

File: src/qutip_cupy/dense.py
Comment (hodgestar) [reply src/qutip_cupy/dense.py:161]: Now we are both confused. :)

In this PR, you register specialisations for `diag`, `identity` and `zeros` -- so those are already used.

You don't register specialisations for `empty` and `emtpy_like` so those are unused (and hence I am suggesting not including them in the qutip-cupy code until they can be used).

File: src/qutip_cupy/__init__.py
Comment (MrRobot2211) [reply src/qutip_cupy/__init__.py:52]: Although if we were to apply the same principle as on PR #15 , we should keep`diags` because a later developer might be confused by the name chages between `qutip`  and `qutip-cupy`

**Comment by MrRobot2211** [comment]
> I know this is still in progress, but I did a partial review of what is here to give some earlier feedback and hopefully save some time later.
> 
> Broader comments:
> 
> * We need a lot more tests (probably a few for each new specialization function)
> * The implementation of `diags` looks very odd in places, but I assume it's a work in progress.

Thank you @hodgestar.
Yes, my next step was to bring tests from `dev.major`.

On the `diags` function there are some minimal changes to work with `cupy arrays` instead of `numpy arrays`, but most of the code is borrowedt from `dev.major`. In any case, as it is not being tested I wouldn't put my hands on the fire for this one.

File: src/qutip_cupy/dense.py
Comment (MrRobot2211) [reply src/qutip_cupy/dense.py:161]: Ah, I see, you are riight the thing is they are not dispathced in `dev.major` either,  see [here](https://github.com/qutip/qutip/blob/dev.major/qutip/core/data/constant.py). I think they are for internal use, maybe I should ask @jakelishman .

Comment (jakelishman) [reply src/qutip_cupy/dense.py:161]: Not every function is meant to have a dispatcher - `csr.sort` won't either, because it only makes sense for CSR.  `empty` and `empty_like` might not make sense for CuPy, but for `Dense` and `CSR` they're used heavily in internal Cython code to implement various features such as `copy` and as the allocation base for `add`, and so on.  I don't see any particular use for them in Python-only space, though, so there's no dispatcher.

Comment (hodgestar) [reply src/qutip_cupy/dense.py:161]: The use within `qutip-cupy` would be to create an empty matrix on the GPU without having to copy it from the CPU, but I think we can live without empty for the moment.

File: src/qutip_cupy/__init__.py
Comment (jakelishman) [reply src/qutip_cupy/__init__.py:52]: It's `diags` rather than `diag` because that matched `qutip.qdiags`, and it can accept a list of diagonals rather than just one (as opposed to `np.diag`).  I was mostly just matching pre-existing stuff in QuTiP, but in general I think `diags` makes sense.  I commented on the historical reasons for the naming convention somewhere else, but the gist was that originally I assumed that anything defined in `csr.pyx` or `dense.pyx` would only ever be accessed by qualified access through that module, and having the additional type information was superfluous, consequently adding noise to the code when you're reading it.  It would be clear, say, that you needed a type suffix on a `norm.l2_*` function, though, because those don't have the type in the module name.

From the dispatcher user's perspective, it really doesn't matter what they're called, because they never see the raw objects anyway.  I'm not against renaming things like `csr.diags` to `diags_csr`, and having data's `__init__.pxd` and `__init__.py` import those select names from `CSR`.

File: src/qutip_cupy/dense.py
Comment (jakelishman) [reply src/qutip_cupy/dense.py:161]: It'll never make sense for TensorFlow or any potential symbolic data type, so I wouldn't think it'll ever become a dispatcher. I don't see a reason why you'd ever want the low-level object-creation routines to be abstractly dispatched.  How would you go about adding data in in the correct form if you don't know what type of object you've just created?  That's why the medium-level `diags`, `identity`, `one_element`, etc exist, and those guys do have dispatchers.

The expectation is that if you ever wanted to call the super low-level `csr` functions from Python space, it would be in something that looked a little like:
```python
out = csr.copy_structure(other)
out_scipy = out.as_scipy(full=True)
out_scipy.data[:] = ...
return out
```
where you're returning a concrete type, not an abstract one.

File: src/qutip_cupy/dense.py
Comment (jakelishman) [reply src/qutip_cupy/dense.py:324]: I would say either commit to the code style, or disable that warning globally.  Personally I don't see any problem with the extra spacing - it's better to be consistent, and have a tool you can use to say "this is always the right way to do it" than to say "this is mostly how we do it, but then there's a bunch of exceptions".  In the latter case, you just end up with a higgledy piggledy mess, and you spend a bunch of time arguing in code review with contributors that "this patch is formatted really badly, please fix", and they say "but _your_ code has all sorts of exceptions".  Better to just be hard-nosed about it and say, "it might not be universally perfect, but consistent and unalienable is better in the long run".

**Comment by MrRobot2211** [comment]
I am migrating all test to qutip style because they are thorough, and because it really makes the coding faster.
I am a little bit uncomfortable with some of the Exception raised checks though. 
For example there are checks for the shapes in `add` and `sub` but not in `matmul`

**Comment by MrRobot2211** [comment]
To make `transpose` pass QuTiP's tests I have to call `.item()` (see 
567b944) this is because `cp.trace` returns a [1,1] ` CuPyArray . this forces comunication between CPU and GPU and affects the timings substatially. This same issue will arise in the implementation of `expect` also, maybe we will have to do something similar to what Asier has done to allow `mul` by [1,1] arrays.

**Comment by MrRobot2211** [comment]
Important note there is no test for the  reshape dispatch in QuTiP, so I will create aPR to add it so I can just inherit it here as I did with the rest of the mathematical functions

File: tests/test_dense.py
Comment (hodgestar) [inline tests/test_dense.py:58]: These imports should just be at the top of the test module.

**Comment by MrRobot2211** [comment]
@hodgestar I think there is no harm on leaving the original dense_tests as they are checking the functions firectly on the class and not on the dispatched specializations.

File: src/qutip_cupy/dense_functions.py
Comment (hodgestar) [inline src/qutip_cupy/dense_functions.py:15]: I think here `cp_arr` will be a `CuPyDense` already, so this function should be `CuPyDense._raw_cupy_constructor(cupy.reshape(cp_arr._cp, (n_rows_out, n_cols_out))`?

File: tests/test_dense.py
Comment (hodgestar) [inline tests/test_dense.py:13]:
```suggestion
from qutip_cupy import CuPyDense
from qutip_cupy import dense

# import qutip after qutip_cupy to supress the matplotlib import warning
from qutip.core import data
```
I think importing qutip after qutip_cupy will avoid the warning and save copying and pasting the warning filter in all the tests.

File: tests/test_dense.py
Comment (hodgestar) [inline tests/test_dense.py:71]:
```suggestion
    cpdense_over_2 = cup_arr / 2.0
    qtpdense_over_2 = data.Dense(array) / 2.0

    np.testing.assert_array_equal(cpdense_over_2.to_array(), qtpdense_over_2.to_array())
```
Fix copy and pasted names.

File: tests/test_dense.py
Comment (hodgestar) [inline tests/test_dense.py:81]: Please fix copy and pasted variable names.

It's probably also a good idea to instead or also test
```python
cpdense = CuPyDense(array)
cpdense /= 2
```
to ensure that the inplace operator works as expected.

File: tests/test_dense.py
Comment (hodgestar) [inline tests/test_dense.py:91]: Fix c&p variable names.

File: tests/test_dense.py
Comment (hodgestar) [inline tests/test_dense.py:101]: Fix c&p variable names and the name of the test. Also test `CuPyDense(array) * (2.0 + 1.0j)`.

File: tests/test_mathematics.py
Comment (hodgestar) [inline tests/test_mathematics.py:13]: Same comment about moving the qutip import after importing qutip_cupy and removing the `# noqa`s and warning filter.

**Review by hodgestar — CHANGES_REQUESTED** [review]
Reviewed!

I left quite a few comments, but all of them are requests for relatively small changes, so hopefully they can be implemented or discussed quickly.

File: src/qutip_cupy/dense.py
Comment (MrRobot2211) [reply src/qutip_cupy/dense.py]: yes and already done

File: src/qutip_cupy/dense.py
Comment (MrRobot2211) [reply src/qutip_cupy/dense.py]: changed and the tests are being habdled by qutip

File: tests/test_mathematics.py
Comment (MrRobot2211) [reply tests/test_mathematics.py:13]: After checking we do not need this filter in particular, and there is no noqa comment needed then. 
But @hodgestar , from past comments , is the correct order base python libraries, external libraries, our own modules?

Comment (hodgestar) [reply tests/test_mathematics.py]: By default, yes -- Python standard library, then third party packages, then this package. However, sometimes exceptions can be made if needed in strange cases such as this, which is why I suggested keeping a comment.

Note that if you run all the tests with `pytest tests` then only the first test file would need the strange ordering but I recommend keeping the strange ordering in all test files so that they can also be run individually with `pytest tests/test_foo.py`.

File: src/qutip_cupy/dense.py
Comment (MrRobot2211) [reply src/qutip_cupy/dense.py:368]: The check is required because the tests in qutip look for a specific exception. If you remove it all incorrect_shape tests fail

Comment (MrRobot2211) [reply src/qutip_cupy/dense.py:368]: Besides I think that if you do not check it the default behaviour in numpy and cupy is to broadcast the smallest onto the largest.

File: src/qutip_cupy/dense_functions.py
Comment (MrRobot2211) [reply src/qutip_cupy/dense_functions.py:15]: yes, good catch

File: src/qutip_cupy/dense_functions.py
Comment (MrRobot2211) [reply src/qutip_cupy/dense_functions.py]: Fixed and turned into docstring

File: src/qutip_cupy/__init__.py
Comment (MrRobot2211) [reply src/qutip_cupy/__init__.py:26]: They should all be E402. You are right.

File: src/qutip_cupy/dense.py
Comment (MrRobot2211) [reply src/qutip_cupy/dense.py]: Good catch.  I reordered the if though.

File: src/qutip_cupy/dense.py
Comment (MrRobot2211) [reply src/qutip_cupy/dense.py:101]: No good reason. I am changing it.

**Comment by MrRobot2211** [comment]
@hodgestar  I think I took on all the changes, except this one https://github.com/qutip/qutip-cupy/pull/10#discussion_r677497874 for which I still need need your feedback.

Some of your suggestions I could not apply directly because the caused breaking changes. Tell me if my changes are to your liking.  

After this one there are 4 very shorts PRs ready.

File: src/qutip_cupy/dense.py
Comment (hodgestar) [reply src/qutip_cupy/dense.py:367]: Re broadcasting: Good point. Let's keep the checks!

**Review by hodgestar — APPROVED** [review]
@MrRobot2211 Thank you for making all the changes! Looks great.

Merge when you're ready.

**Pull request merged**
